File: src/test/java/com/wanted/preonboarding/ticket/contoller/ReservationControllerTest.java
package com.wanted.preonboarding.ticket.contoller;

import static com.wanted.preonboarding.ticket.exception.ExceptionMessage.NOT_FOUND_PERFORMANCE;
import static com.wanted.preonboarding.ticket.exception.ExceptionMessage.NOT_FOUND_SEAT_INFO;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.BDDMockito.given;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.wanted.preonboarding.ticket.domain.entity.Performance;
import com.wanted.preonboarding.ticket.domain.entity.ReservationStatus;
import com.wanted.preonboarding.ticket.exception.EntityNotFound;
import com.wanted.preonboarding.ticket.exception.InsufficientPaymentException;
import com.wanted.preonboarding.ticket.service.ReservationService;
import com.wanted.preonboarding.ticket.service.dto.request.ReservationCheckRequestDto;
import com.wanted.preonboarding.ticket.service.dto.request.ReservationRequestDto;
import com.wanted.preonboarding.ticket.service.dto.response.ReservationCheckResponseDto;
import com.wanted.preonboarding.ticket.service.dto.response.ReservationResponseDto;

@WebMvcTest(controllers = ReservationController.class)
class ReservationControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @MockBean
    private ReservationService reservationService;

    @Test
    void 예약_성공() throws Exception {
        // given
        final Performance performance = Performance.builder()
                .id(UUID.fromString("4438a3e6-b01c-11ee-9426-0242ac180002"))
                .name("공연 이름")
                .price(10000)
                .status(ReservationStatus.AVAILABLE)
                .start_date(LocalDateTime.of(2024, 2, 1, 10, 0))
                .build();

        final ReservationResponseDto responseDto = ReservationResponseDto.builder()
                .performanceId(performance.getId())
                .performanceName(performance.getName())
                .reservationName("홍길동")
                .reservationPhoneNumber("010-1234-5678")
                .amount(0)
                .round(1)
                .line('A')
                .seat(1)
                .build();

        given(reservationService.reserve(any()))
                .willReturn(responseDto);

        // when
        // then
        final ReservationRequestDto requestDto = ReservationRequestDto.builder()
                .reservationName("홍길동")
                .reservationPhoneNumber("010-1234-5678")
                .performanceId(UUID.fromString("4438a3e6-b01c-11ee-9426-0242ac180002"))
                .amount(10000)
                .round(1)
                .line('A')
                .seat(1)
                .build();

        mockMvc.perform(post("/reserve")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(requestDto)))
                .andExpect(status().isCreated())
                .andExpectAll(
                        jsonPath("$.data.performanceId").value(performance.getId().toString()),
                        jsonPath("$.data.performanceName").value(performance.getName()),
                        jsonPath("$.data.reservationName").value("홍길동"),
                        jsonPath("$.data.reservationPhoneNumber").value("010-1234-5678"),
                        jsonPath("$.data.amount").value(0),
                        jsonPath("$.data.round").value(1),
                        jsonPath("$.data.line").value("A"),
                        jsonPath("$.data.seat").value(1))
                .andDo(print());
    }

    @Test
    void 공연이_존재하지_않으면_예약_실패() throws Exception {
        // given
        given(reservationService.reserve(any()))
                .willThrow(new EntityNotFound(NOT_FOUND_PERFORMANCE.getMessage()));

        // when
        // then
        final ReservationRequestDto requestDto = ReservationRequestDto.builder()
                .reservationName("홍길동")
                .reservationPhoneNumber("010-1234-5678")
                .performanceId(UUID.fromString("4438a3e6-b01c-11ee-9426-0242ac180002"))
                .amount(10000)
                .round(1)
                .line('B')
                .seat(1)
                .build();

        mockMvc.perform(post("/reserve")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(requestDto)))
                .andExpect(status().isNotFound())
                .andDo(print());
    }

    @Test
    void 좌석이_존재하지_않으면_예약_실패() throws Exception {
        // given
        given(reservationService.reserve(any()))
                .willThrow(new EntityNotFound(NOT_FOUND_SEAT_INFO.getMessage()));

        // when
        // then
        final ReservationRequestDto requestDto = ReservationRequestDto.builder()
                .reservationName("홍길동")
                .reservationPhoneNumber("010-1234-5678")
                .performanceId(UUID.fromString("4438a3e6-b01c-11ee-9426-0242ac180002"))
                .amount(10000)
                .round(1)
                .line('B')
                .seat(1)
                .build();

        mockMvc.perform(post("/reserve")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(requestDto)))
                .andExpect(status().isNotFound())
                .andDo(print());
    }

    @ParameterizedTest
    @ValueSource(ints = {9999, 9000, 6000, 4000})
    void 잔고가_부족하면_예약_실패(final int amount) throws Exception {
        // given
        final Performance performance = Performance.builder()
                .id(UUID.fromString("4438a3e6-b01c-11ee-9426-0242ac180002"))
                .name("공연 이름")
                .price(10000)
                .status(ReservationStatus.AVAILABLE)
                .start_date(LocalDateTime.of(2024, 2, 1, 10, 0))
                .build();

        given(reservationService.reserve(
                        argThat(requestDto -> requestDto.amount() < performance.getPrice())))
                .willThrow(InsufficientPaymentException.class);

        // when
        // then
        final ReservationRequestDto requestDto = ReservationRequestDto.builder()
                .reservationName("홍길동")
                .reservationPhoneNumber("010-1234-5678")
                .performanceId(UUID.fromString("4438a3e6-b01c-11ee-9426-0242ac180002"))
                .amount(amount)
                .round(1)
                .line('B')
                .seat(1)
                .build();

        mockMvc.perform(post("/reserve")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(requestDto)))
                .andExpect(status().isBadRequest())
                .andDo(print());
    }

    @Test
    void 예약_조회() throws Exception {
        // given
        given(reservationService.check(any()))
                .willReturn(List.of(ReservationCheckResponseDto.builder()
                        .reservationName("홍길동")
                        .reservationPhoneNumber("010-1234-5678")
                        .performanceId(UUID.fromString("4438a3e6-b01c-11ee-9426-0242ac180002"))
                        .performanceName("공연 이름")
                        .round(1)
                        .line('A')
                        .seat(1)
                        .build()));

        // when
        // then
        final ReservationCheckRequestDto request = ReservationCheckRequestDto.builder()
                .reservationName("홍길동")
                .reservationPhoneNumber("010-1234-5678")
                .build();
        mockMvc.perform(get("/reserve")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isOk())
                .andDo(print());
    }
}

File: src/test/kotlin/com/wanted/preonboarding/ticket/presentation/ReservationControllerTest.kt
package com.wanted.preonboarding.ticket.presentation

import com.jayway.jsonpath.JsonPath
import com.wanted.preonboarding.ticket.domain.PerformanceType
import com.wanted.preonboarding.ticket.domain.SeatInfoFixtureBuilder
import com.wanted.preonboarding.ticket.domain.UserInfoFixtureBuilder
import com.wanted.preonboarding.ticket.infra.entity.PerformanceEntity
import com.wanted.preonboarding.ticket.infra.entity.PerformanceSeatInfoEntity
import com.wanted.preonboarding.ticket.infra.entity.ReservationEntity
import com.wanted.preonboarding.ticket.presentation.request.ReservationCancelRequest
import com.wanted.preonboarding.ticket.presentation.request.ReservationRequest
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test
import org.springframework.http.HttpMethod
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.LocalDateTime
import java.util.UUID

class ReservationControllerTest : IntegrationTest() {
    @Test
    fun `공연이 존재하지 않으면 예약에 실패한다`() {
        // given
        val request = ReservationRequest(
            performanceId = UUID.randomUUID(),
            userInfo = UserInfoFixtureBuilder().build(),
            balance = 10000,
            seatInfo = SeatInfoFixtureBuilder().build(),
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "존재하지 않는 공연입니다."
        }
    }

    @Test
    fun `예약이 마감된 공연은 예약할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = false,
            )
        )
        val request = ReservationRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            balance = 10000,
            seatInfo = SeatInfoFixtureBuilder().build(),
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "예약이 마감되었습니다."
        }
    }

    @Test
    fun `잔액이 부족하면 예약할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val request = ReservationRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            balance = 10,
            seatInfo = SeatInfoFixtureBuilder().build(),
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "잔액이 부족합니다."
        }
    }

    @Test
    fun `좌석이 존재하지 않으면 예약할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val request = ReservationRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            balance = 10000,
            seatInfo = SeatInfoFixtureBuilder().build(),
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "존재하지 않는 좌석입니다."
        }
    }

    @Test
    fun `예약할 수 없는 좌석은 예약할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val seatInfo = SeatInfoFixtureBuilder().build()
        save(
            PerformanceSeatInfoEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                isReserve = false,
            )
        )
        val request = ReservationRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            balance = 10000,
            seatInfo = seatInfo,
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "존재하지 않는 좌석입니다."
        }
    }

    @Test
    fun `이미 예약된 좌석은 예약할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val seatInfo = SeatInfoFixtureBuilder().build()
        save(
            PerformanceSeatInfoEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                isReserve = true,
            )
        )
        save(
            ReservationEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                name = "name",
                phoneNumber = "01012345678",
            )
        )
        val request = ReservationRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            balance = 10000,
            seatInfo = seatInfo,
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "이미 예약된 좌석입니다."
        }
    }

    @Test
    fun `예약에 성공한다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val seatInfo = SeatInfoFixtureBuilder().build()
        save(
            PerformanceSeatInfoEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                isReserve = true,
            )
        )
        val request = ReservationRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            balance = 10000,
            seatInfo = seatInfo,
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations",
            body = request,
        )

        // then
        response.andExpect(status().isOk()).andExpect {
            val performanceId = JsonPath.read(it.response.contentAsString, "$.data.performanceId") as String
            val name = JsonPath.read(it.response.contentAsString, "$.data.userInfo.name") as String
            val phoneNumber = JsonPath.read(it.response.contentAsString, "$.data.userInfo.phoneNumber") as String
            val round = JsonPath.read(it.response.contentAsString, "$.data.seatInfo.round") as Int
            val gate = JsonPath.read(it.response.contentAsString, "$.data.seatInfo.gate") as Int
            val line = JsonPath.read(it.response.contentAsString, "$.data.seatInfo.line") as String
            val seat = JsonPath.read(it.response.contentAsString, "$.data.seatInfo.seat") as Int

            performanceId shouldBe performance.id.toString()
            name shouldBe request.userInfo.name
            phoneNumber shouldBe request.userInfo.phoneNumber
            round shouldBe seatInfo.round
            gate shouldBe seatInfo.gate
            line shouldBe seatInfo.line
            seat shouldBe seatInfo.seat

            val reservation = findOne<ReservationEntity>(
                """
                    SELECT r FROM ReservationEntity r
                    WHERE r.performanceId = :performanceId AND r.name = :name AND r.phoneNumber = :phoneNumber
                """.trimIndent(),
                mapOf(
                    "performanceId" to performance.id!!,
                    "name" to request.userInfo.name,
                    "phoneNumber" to request.userInfo.phoneNumber,
                )
            )
            reservation.performanceId shouldBe performance.id
            reservation.name shouldBe request.userInfo.name
            reservation.phoneNumber shouldBe request.userInfo.phoneNumber
            reservation.round shouldBe seatInfo.round
            reservation.gate shouldBe seatInfo.gate
            reservation.line shouldBe seatInfo.line
            reservation.seat shouldBe seatInfo.seat

            val performanceSeatInfo = findOne<PerformanceSeatInfoEntity>(
                """
                    SELECT psi FROM PerformanceSeatInfoEntity psi
                    WHERE psi.performanceId = :performanceId AND psi.round = :round AND psi.gate = :gate AND psi.line = :line AND psi.seat = :seat
                """.trimIndent(),
                mapOf(
                    "performanceId" to performance.id!!,
                    "round" to seatInfo.round,
                    "gate" to seatInfo.gate,
                    "line" to seatInfo.line,
                    "seat" to seatInfo.seat,
                )
            )
            performanceSeatInfo.isReserve shouldBe false
        }
    }

    @Test
    fun `공연이 존재하지 않으면 예약취소에 실패한다`() {
        // given
        val request = ReservationCancelRequest(
            performanceId = UUID.randomUUID(),
            userInfo = UserInfoFixtureBuilder().build(),
            seatInfo = SeatInfoFixtureBuilder().build(),
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations/cancel",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "존재하지 않는 공연입니다."
        }
    }

    @Test
    fun `좌석이 존재하지 않으면 예약을 취소할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val request = ReservationCancelRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            seatInfo = SeatInfoFixtureBuilder().build(),
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations/cancel",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "존재하지 않는 좌석입니다."
        }
    }

    @Test
    fun `예약된 내역이 존재하지 않으면 예약을 취소할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val seatInfo = SeatInfoFixtureBuilder().build()
        save(
            PerformanceSeatInfoEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                isReserve = true,
            )
        )
        val request = ReservationCancelRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            seatInfo = seatInfo
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations/cancel",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "예약된 내역이 없습니다."
        }
    }

    @Test
    fun `타인이 한 예약은 취소할 수 없다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val seatInfo = SeatInfoFixtureBuilder().build()
        save(
            PerformanceSeatInfoEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                isReserve = true,
            )
        )
        save(
            ReservationEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                name = "모르는 사람",
                phoneNumber = "01012345678",
            )
        )
        val request = ReservationCancelRequest(
            performanceId = performance.id!!,
            userInfo = UserInfoFixtureBuilder().build(),
            seatInfo = seatInfo,
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations/cancel",
            body = request,
        )

        // then
        response.andExpect {
            val message = JsonPath.read(it.response.contentAsString, "$.error") as String
            message shouldBe "예약된 내역이 없습니다."
        }
    }

    @Test
    fun `예약 취소에 성공한다`() {
        // given
        val performance = save(
            PerformanceEntity(
                name = "performance",
                type = PerformanceType.CONCERT,
                price = 10000,
                round = 1,
                startDate = LocalDateTime.now(),
                isReserve = true,
            )
        )
        val seatInfo = SeatInfoFixtureBuilder().build()
        val userInfo = UserInfoFixtureBuilder().build()
        save(
            PerformanceSeatInfoEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                isReserve = false,
            )
        )
        save(
            ReservationEntity(
                performanceId = performance.id!!,
                round = seatInfo.round,
                gate = seatInfo.gate,
                line = seatInfo.line,
                seat = seatInfo.seat,
                name = userInfo.name,
                phoneNumber = userInfo.phoneNumber
            )
        )
        val request = ReservationCancelRequest(
            performanceId = performance.id!!,
            userInfo = userInfo,
            seatInfo = seatInfo,
        )

        // when
        val response = request(
            method = HttpMethod.POST,
            url = "/v1/reservations/cancel",
            body = request,
        )

        // then
        response.andExpect(status().isOk()).andExpect {
            val reservation = findAll<ReservationEntity>(
                """
                    SELECT r FROM ReservationEntity r
                    WHERE r.performanceId = :performanceId AND r.name = :name AND r.phoneNumber = :phoneNumber
                """.trimIndent(),
                mapOf(
                    "performanceId" to performance.id!!,
                    "name" to userInfo.name,
                    "phoneNumber" to userInfo.phoneNumber,
                )
            )
            reservation.size shouldBe 0

            val performanceSeatInfo = findOne<PerformanceSeatInfoEntity>(
                """
                    SELECT psi FROM PerformanceSeatInfoEntity psi
                    WHERE psi.performanceId = :performanceId AND psi.round = :round AND psi.gate = :gate AND psi.line = :line AND psi.seat = :seat
                """.trimIndent(),
                mapOf(
                    "performanceId" to performance.id!!,
                    "round" to seatInfo.round,
                    "gate" to seatInfo.gate,
                    "line" to seatInfo.line,
                    "seat" to seatInfo.seat,
                )
            )
            performanceSeatInfo.isReserve shouldBe true
        }
    }
}
